### ISSUE TITLE
The Editor fails to launch if it was previously closed while still processing assets for the project with all gems enabled

### ISSUE THREAD

**Issue by LB-TomaszSerwanski** [o3de/o3de#11926]
**Describe the bug**
The Editor crashes on the splash screen if the User has previously closed it while still processing the assets. Please note that the issue only occurs on projects with all gems enabled.

Please note that the issue does not always occur is most likely to reproduce by closing the Editor while there is still over 800 jobs to be processed in the Asset Processor.

Additionally - we have noticed that closing the Editor on project with all gems enabled when still processing assets also closes the Asset Processor. This is not the case when doing so on the default template project, on which the Asset Processor remains launched and keeps processing the assets.

The workaround for the issue is to delete the cache folder of the project and reprocess all the assets.

Please refer to the attached video and log files for more details.

**Call Stack Trace**
```
 9) 00007FFD3ED70057 (KERNELBASE) : UnhandledExceptionFilter
 8) 00007FFD3E1ED6D8 (InProcessClient64) : (function-name not available)
 7) 00007FFD3E1ED321 (InProcessClient64) : (function-name not available)
 6) 00007FFD40F753B0 (ntdll) : memset
 5) 00007FFD40F5C766 (ntdll) : _C_specific_handler
 4) 00007FFD40F7229F (ntdll) : _chkstk
 3) 00007FFD40F21454 (ntdll) : RtlRaiseException
 2) 00007FFD40F70DCE (ntdll) : KiUserExceptionDispatcher
 1) D:\Git\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Model\Model.cpp (62) : AZ::RPI::Model::GetLods
```

**Steps to reproduce**
1. Create a new project with all gems enabled and build it.
2. Launch the Editor for the project.
3. Close the Editor while the Asset Processor is still working on the assets (ideally when there is ~850 jobs left).
4. Relaunch the Editor.

**Expected behavior**
The Editor launches without issues.

**Actual behavior**
The Editor crashes on the splash screen.

**Screenshots/Video**

https://user-images.githubusercontent.com/86953108/189912846-a02bb6b0-2562-4ff0-bd29-9bfdfe15b6c6.mp4

**Found in Branch**
Development (511a88c)
Stabilization/2210 (021d32a)

**Desktop/Device:**
 - Device: PC
 - OS: Windows
 - Version 10
 - CPU AMD Ryzen 5 3600
 - GPU Nvidia RTX 2060 SUPER
 - Memory 16GB

**Attachments**
[Editor.log](https://github.com/o3de/o3de/files/9557514/Editor.log)
[error.log](https://github.com/o3de/o3de/files/9557515/error.log)
[error.zip](https://github.com/o3de/o3de/files/9557516/error.zip)

**Comment by monroegm** [comment]
@o3de/sig-graphics-audio Based on the error message, do you think this is a graphics issue?

**Comment by jackalbe** [comment]
```
        void DiffuseProbeGridFeatureProcessor::OnVisualizationModelAssetReady(Data::Asset<Data::AssetData> asset)
        {
            Data::Asset<RPI::ModelAsset> modelAsset = asset;

            m_visualizationModel = RPI::Model::FindOrCreate(modelAsset);
            AZ_Assert(m_visualizationModel.get(), "Failed to load DiffuseProbeGrid visualization model");

            const AZStd::span<const Data::Instance<RPI::ModelLod>>& modelLods = m_visualizationModel->GetLods();
            AZ_Assert(!modelLods.empty(), "Invalid DiffuseProbeGrid visualization model");

```
There is an AZ_Assert but then the code continues until it hits the m_visualizationModel nullptr. This will result in an instant crash outside the debugger. Right?

**Comment by dmcdiarmid-ly** [comment]
The visualization model is a critical asset, which is why the code is written that way.  The model must be there when the Editor starts.

Based on these AssetManager warnings, it appears there is a problem with the model in the build:

[Warning] (AssetContainer) - Error loading asset {94A0878A-DBD6-5B0E-994E-BEB870442F9F}:0[Warning] (AssetContainer) - Error loading asset {96CFA695-08D7-59EB-AA3E-8BE645466F23}:0[Warning] (AssetContainer) - Asset models/diffuseprobesphere.azmodel ({8CC9B575-39D7-5914-AC42-92A77833E63A}:1003ea43) references/depends on asset {0010FB8C-0325-5565-9FE6-D24A75401BF7}:0 which does not exist in the catalog and cannot be loaded.[Warning] (AssetManager) - GetAsset called for asset which does not exist in asset catalog and cannot be loaded. Asset may be missing, not processed or moved. AssetId: {0010FB8C-0325-5565-9FE6-D24A75401BF7}:0[Error] (RPI::Buffer) - Failed to acquire the buffer pool instance from asset.(System)

**Comment by amzn-mike** [comment]
The resource builder is not marking its jobs as critical, that would explain why these aren't ready:

```c++
void ResourcePoolBuilder::CreateJobs(const AssetBuilderSDK::CreateJobsRequest& request, AssetBuilderSDK::CreateJobsResponse& response)
        {
            if (m_isShuttingDown)
            {
                response.m_result = AssetBuilderSDK::CreateJobsResultCode::ShuttingDown;
                return;
            }

            for (const AssetBuilderSDK::PlatformInfo& platformInfo : request.m_enabledPlatforms)
            {
                AssetBuilderSDK::JobDescriptor descriptor;
                descriptor.m_jobKey = "Atom Resource Pool";
                descriptor.SetPlatformIdentifier(platformInfo.m_identifier.c_str());
                descriptor.m_critical = false;

                response.m_createJobOutputs.push_back(descriptor);
            }

            response.m_result = AssetBuilderSDK::CreateJobsResultCode::Success;
        }
```

**Comment by tonybalandiuk** [comment]
per feedback from @galibzon "This is an Atom issue with assets not being marked as critical. The Atom ResourcePoolBuilder needs to be updated. We could probably mark this is a known issue as there is an "easy" workaround (ie don't close editor while assets are processed first time)."

**Comment by galibzon** [comment]
@VickyAtAZ , can you please mark resourcepool assets as critical and see if it fixes the issue?

**Comment by tonybalandiuk** [comment]
decision has been made to attempt to fix this in stabilization since this could result in a poor first time user experience

**Comment by galibzon** [comment]
I found out why the new crash occurs. SimplePointLightFeatureProcessor::Activate() is being called BEFORE RPISystem::InitializeSystemAssets() is called.

**Comment by galibzon** [comment]
Here are some scenarios I have tested after setting `descriptor.m_critical = true;` in ResourcePoolBuilder:
Case 1:

**Comment by VickyAtAZ** [comment]
I think we may need to address this issue from editor initialization. 
This the behavior I have seen:
1. Start editor with clean cache, the editor would wait for all asset jobs to be finished before it move to show the editor window. And it would initialize successfully. 
2.If user terminates editor before it fully started (AP still processing file), the AP is terminated as well. The second time when editor starts, it won't wait for AP to process the pending jobs which could lead to any random issue due to unprocessed assets. This may even include critical assets.   

I think we need to find out why the Engine/Editor behave differently between the first run and second run.

**Comment by galibzon** [comment]
@amzn-mike To provide feedback.

**Comment by VickyAtAZ** [comment]
I can confirm that the editor won't even wait for finish processing critical assets when it runs the second time.

**Comment by amzn-mike** [comment]
I'm looking into this currently.  It looks like a minor issue with how CompileAssetSync is being called in a few places and a bigger issue of editor starting up without having an assetcatalog.xml on disk due to AP being interrupted the first time.

So I think we generally don't see this after first startup because the assetcatalog.xml would have been written out.   But if AP is interrupted on its first run, it can complete all the "critical" assets before its 100% done, and it doesn't seem to write out the catalog until hitting that 100% done state.  That means next launch, AP is immediately "ready" but still not done, so the editor moves past the waiting for AP stage.

I'm still looking into what's going on with the catalog.

**Comment by amzn-mike** [comment]
It looks like the issue is the catalog is not being rebuilt for the editor on the 2nd run.  I'm looking into having the AP send the updates to the editor when it rebuilds its own catalog

**Comment by amzn-mike** [comment]
I have a potential fix, just need to do a bit more testing: https://github.com/o3de/o3de/pull/12445

**Comment by lmbr-pip** [comment]
@tonybalandiuk Suggesting that we remove this issue from release consideration. There are some AR instabilities in stabilization with the suggested fix that need looking into. Bug is long standing behaviour and not new, workaround is to clear cache and try again, or just launch asset processor and let it finish, if you encounter this issue.

We will continue to work on fix, including some of the performance orientated suggestions from the PR, and deliver to development so it could be cherry picked to patch if needed.

**Comment by tonybalandiuk** [comment]
@chanmosq This is the one the team was just working on but we didn't resolve it. Ok to release note this, then remove it from the 2210 project. Thanks.

**Comment by chanmosq** [comment]
Added to the release notes: 
> - For projects with all Gems enabled, if a user previously closed the O3DE Editor while it was still processing assets, the O3DE Editor crashes on splash screen when it's reopened. [#11926](https://github.com/o3de/o3de/issues/11926) This issue will be resolved in development: [#12445](https://github.com/o3de/o3de/pull/12445)

**Comment by rainbj** [comment]
I do want to point out that this use case of "Add all Gems" is a very unusual use case. From a UX perspective, I don't imagine we want most people to do this. But I do understand the need to figure out this issue.

**Comment by bhanuja-s** [comment]
> I do want to point out that this use case of "Add all Gems" is a very unusual use case.

I agree, from previous research (before O3DE's Project Manager) we found that users would enable all Gems to cut corners with build and compilation time that following enabling Gems over a period of time. We've decisively reduced this time through the new onboarding workflow.

**Comment by AMZN-stankowi** [comment]
@galibzon this issue has not been completed yet, the Editor still crashes on launch. I just posted a pull request here https://github.com/o3de/o3de/pull/13565

**Comment by AMZN-stankowi** [comment]
The fix has been merged to development

**Comment by LB-TomaszSerwanski** [comment]
Reopening the issue as it still occurs in the latest Development ([35ee631](https://github.com/o3de/o3de/commit/35ee6319dac8bbfeaac69b045d57ca51ac12f7d7)) branch. We have verified that this also reproduces on the default template project (as well as on the project with all gems enabled). The call stack looks exactly the same as the one attached to the issue on both projects but I am attaching it to this comment just in case (the attached files are from the default template project): 

- [Editor.log](https://github.com/o3de/o3de/files/10217044/Editor.log)
- [error.log](https://github.com/o3de/o3de/files/10217045/error.log)
- [error.zip](https://github.com/o3de/o3de/files/10217047/error.zip)

Please also note that this also happens on the Installer build (as described in #13602).

**Comment by AMZN-stankowi** [comment]
This looks like it's a different crash than the one I fixed, and this should be re-triaged or a new ticket created.
```

==================================================================
(System) - Trace::Assert
 D:/Git/o3de/Gems/DiffuseProbeGrid/Code/Source/Render/DiffuseProbeGridFeatureProcessor.cpp(855): (10272) 'void __cdecl AZ::Render::DiffuseProbeGridFeatureProcessor::OnVisualizationModelAssetReady(class AZ::Data::Asset<class AZ::Data::AssetData>)'
(System) - Failed to load DiffuseProbeGrid visualization model
(System) - ------------------------------------------------
(System) - D:\Git\o3de\Gems\DiffuseProbeGrid\Code\Source\Render\DiffuseProbeGridFeatureProcessor.cpp (857) : AZ::Render::DiffuseProbeGridFeatureProcessor::OnVisualizationModelAssetReady
(System) - D:\Git\o3de\Gems\DiffuseProbeGrid\Code\Source\Render\DiffuseProbeGridFeatureProcessor.cpp (941) : AZ::Render::DiffuseProbeGridFeatureProcessor::OnAssetReady
(System) - D:\Git\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (146) : AZ::Internal::EBusContainer<AZ::Data::AssetEvents,AZ::Data::AssetEvents,1,1>::Dispatcher<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents> >::Event<void (__cdecl AZ::Data::
(System) - D:\Git\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (1851) : AZ::Data::AssetManager::NotifyAssetReady
(System) - D:\Git\o3de\Code\Framework\AzCore\AzCore\std\function\function_template.h (148) : AZStd::Internal::function_util::get_invoker<void __cdecl(void),AZStd::Internal::function_util::function_obj_tag,AZStd::stateless_allocator>::call<`AZ::BusInternal::EBusBroa
(System) - D:\Git\o3de\Code\Framework\AzCore\AzCore\EBus\Policies.h (267) : AZ::EBusQueuePolicy<1,AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZStd::recursive_mutex>::Execute
(System) - D:\Git\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (591) : AZ::Data::AssetManager::DispatchEvents
(System) - D:\Git\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (339) : AZ::Data::WaitForAsset::Wait
(System) - D:\Git\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (323) : AZ::Data::WaitForAsset::WaitUntilReady
(System) - D:\Git\o3de\Code\Framework\AzCore\AzCore\Asset\AssetManager.cpp (849) : AZ::Data::AssetManager::BlockUntilLoadComplete
(System) - D:\Git\o3de\Code\Framework\AzCore\AzCore\Asset\AssetCommon.h (1153) : AZ::Data::Asset<AZ::RPI::ShaderAsset>::BlockUntilLoadComplete
(System) - D:\Git\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\RPIUtils.cpp (575) : AZ::RPI::FindShaderAsset
(System) - D:\Git\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\RPIUtils.cpp (593) : AZ::RPI::LoadShader
(System) - D:\Git\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\RPIUtils.cpp (633) : AZ::RPI::LoadCriticalShader
(System) - D:\Git\o3de\Gems\Atom\Feature\Common\Code\Source\ReflectionProbe\ReflectionProbeFeatureProcessor.cpp (667) : AZ::Render::ReflectionProbeFeatureProcessor::LoadShader
(System) - D:\Git\o3de\Gems\Atom\Feature\Common\Code\Source\ReflectionProbe\ReflectionProbeFeatureProcessor.cpp (62) : AZ::Render::ReflectionProbeFeatureProcessor::Activate
(System) - D:\Git\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Scene.cpp (156) : AZ::RPI::Scene::Activate
(System) - D:\Git\o3de\Gems\Atom\Bootstrap\Code\Source\BootstrapSystemComponent.cpp (361) : AZ::Render::Bootstrap::BootstrapSystemComponent::GetOrCreateAtomSceneFromAzScene
(System) - D:\Git\o3de\Gems\Atom\Tools\AtomToolsFramework\Code\Source\Viewport\RenderViewportWidget.cpp (170) : `AtomToolsFramework::RenderViewportWidget::SetScene'::`2'::<lambda_1>::operator()
(System) - D:\Git\o3de\Code\Framework\AzCore\AzCore\EBus\Internal\BusContainer.h (1541) : AZ::Internal::EBusContainer<AZ::Render::Bootstrap::Request,AZ::Render::Bootstrap::Request,0,0>::Dispatcher<AZ::EBus<AZ::Render::Bootstrap::Request,AZ::Render::Bootstrap::Reque
(System) - D:\Git\o3de\Gems\Atom\Tools\AtomToolsFramework\Code\Source\Viewport\RenderViewportWidget.cpp (179) : AtomToolsFramework::RenderViewportWidget::SetScene
(System) - D:\Git\o3de\Code\Editor\EditorViewportWidget.cpp (2196) : EditorViewportWidget::UpdateScene
(System) - D:\Git\o3de\Code\Editor\EditorViewportWidget.cpp (844) : EditorViewportWidget::SetViewportId
(System) - D:\Git\o3de\Code\Editor\ViewPane.cpp (762) : CLayoutViewPane::AttachViewport
(System) - D:\Git\o3de\Code\Editor\ViewPane.cpp (678) : CLayoutViewPane::SetViewClass

```

**Comment by LB-TomaszSerwanski** [comment]
Hi @AMZN-stankowi, the log that you pasted is also present in the Editor.log file attached to the issue originally (please refer to the attachments section of the bug). The **call stack trace** section from the description contains the log from the Error.log file (and still matches the one attached after reopening the issue). Please confirm if this is still a different crash, and if so, we will log it separately. Thanks!

**Comment by LB-JaroslawGaweda** [comment]
It should be noted that this issue is still being encountered on build Development (c9368cc) installer.

**Comment by LB-BartoszDudzinski** [comment]
We noticed that [#13247](https://github.com/o3de/o3de/issues/13247) was closed as a duplicate of this issue even though was related to the launching the Editor without processed assets for the first time. The issue is slightly different now, as the Editor freezes while the assets are being processed and a several minutes after the AP becomes idle until it becomes responsive on Windows platform. On Linux, the Editor crashes while the assets are still being processed and only after fully processing assets, the Editor can be launched. Considering the similarity in those issues, should we log a separate ticket for this, or will it be tracked here as the source of this issues may be the same?
